### ISSUE TITLE
Default 'escaping' of field output in Twig.

### ISSUE THREAD

**Issue by bobdenotter** [bolt/four#351]
In Bolt 3, if you do `{{ record.fieldname }}`, it'll output HTML. At the same time, it's not *explicitly defined* how it should behave. One of the causes of this is how the fields are stored internally. 

We should make sure that Bolt 4 does this sensibly, consistent and robust. In my opinion, it should work like the following. Let's say we have a title field, with a `<i>` tag in it.

![schermafbeelding 2019-03-07 om 13 59 32](https://user-images.githubusercontent.com/1833361/53958406-4458cb80-40e1-11e9-9c73-87f0dc955248.png)

In Twig we should be able to do: 

```twig
<hr>
    <p>{{ record.title }}</p>
    <p>{{ record.title|raw }}</p>
    <p>{{ record.title|e('html') }}</p>
    <p>{{ dump(record.title) }}</p>
    <p>{{ dump(record.title.getValue()) }}</p>
<hr>
```

And it should output: 

![schermafbeelding 2019-03-07 om 14 07 37](https://user-images.githubusercontent.com/1833361/53958875-66068280-40e2-11e9-9565-c9b91e9a8a40.png)

This way, the _implementor_ can use `{{ record.foo }}` and get the HTML output they expect 90% of the time, and at the same time, you still have access to the underlying data structure to do with it what you need.

**Comment by stryxx** [comment]
@bobdenotter What does happen there when you add to field <script src="../link/to/bitcoin.miner.js"></script>? :)

**Comment by JarJak** [comment]
IMO, since Bolt is a WYSIWYG editor, if an admin puts `<i>Something</i>` in a text or textarea field, he should see `<i>Something</i>` on website. That's because this type of field was not made to store html data but raw text.
On the other hand, if the field has type of html, as admin expects html output on website, so he should see *Something*.

BUT we could add an option to make simple text field behave as html one (e.g. a field to paste some js stuff). Example:
```
ga_script:
    type: text
    safe: true # default: false
```

**Comment by bobdenotter** [comment]
@stryxx Bolt 3 had a whitelist-based filtering for HTML-like fields, where input would be sanitized before going into the database. In Bolt 3's config.yml, you had: 

```yaml
# Define the HTML tags and attributes that are allowed in 'cleaned' HTML. This
# is used for sanitizing HTML, to make sure there are no undesirable elements
# left in the content that is shown to users. For example, tags like `<script>`
# or `onclick`-attributes.
# Note: enabling options in the `wysiwyg` settings will implicitly add items to
# the allowed tags. For example, if you set `images: true`, the `<img>` tag
# will be allowed, regardless of it being in the `allowed_tags` setting.
htmlcleaner:
    allowed_tags: [ div, span, p, br, hr, s, u, strong, em, i, b, li, ul, ol, mark, blockquote, pre, code, tt, h1, h2, h3, h4, h5, h6, dd, dl, dt, table, tbody, thead, tfoot, th, td, tr, a, img, address, abbr, iframe, caption, sub, sup, figure, figcaption ]
    allowed_attributes: [ id, class, style, name, value, href, src, alt, title, width, height, frameborder, allowfullscreen, scrolling, target, colspan, rowspan ]
```

This meant that `<script>` would be filtered out, but `<i>` didn't.

**Comment by JarJak** [comment]
@bobdenotter AFAIR it didn't work well in Bolt 3. I could put some JS in e.g. title field and it got executed.

**Comment by stryxx** [comment]
@bobdenotter iframe in allowed tags? :) Soooo... I can execute js in iframe, or... I can add onclick to all of tags. There exists many ways to execute some harmful code.

**Comment by bobdenotter** [comment]
> AFAIR it didn't work well in Bolt 3. I could put some JS in e.g. title field and it got executed in admin area.

Not going to name names, but _someone_ initially only implemented that scrubber only for `type:html` fields, and not for `type: text` fields. This has since been fixed. 

![schermafbeelding 2019-03-07 om 14 25 54](https://user-images.githubusercontent.com/1833361/53959876-1f665780-40e5-11e9-99bf-960bd913b08c.png)

> BUT we could add an option to make simple text field behave as html one (e.g. a field to paste some js stuff). Example:

@JarJak I think I agree that this is a good idea. Two minor concerns though.. 

1) Is `safe:` the best name for this? imho, "safe" is either a desired aspect of something, or something you attribute to it. So, it could be perceived as "This is safe to use!". Maybe something like `plaintext: true` or `allows_markup: true` would be more expressive.

2) i think `type: html` should default to allow HTML, while `type:text` and `type: textarea` shouldn't. Won't it be confusing for users to have different defaults?

**Comment by bobdenotter** [comment]
> iframe in allowed tags? 

@stryxx Yes. it's how virtually all embeds work these days. Contrary to what feels intuitive, `<iframes>` are very much sandboxed in all browsers, making them pretty much harmless. :-)

**Comment by bobdenotter** [comment]
@stryxx In Bolt 3 it is stripped out, and I think we should also strip these out in Bolt 4. IMHO, the correct location for scripts is in the (Twig) templates, and it has no place in _content_.

**Comment by JarJak** [comment]
@bobdenotter it was properly escaped on website, but not in __admin__ area. Try to see listing with this record, you should see an alert there.

> 1. Is safe: the best name for this? imho, "safe" is either a desired aspect of something, or something you attribute to it. So, it could be perceived as "This is safe to use!". Maybe something like plaintext: true or allows_markup: true would be more expressive.

Sure, we can think of a better name for it :)

> 2. i think type: html should default to allow HTML, while type:text and type: textarea shouldn't. Won't it be confusing for users to have different defaults?

They are different fields so even option names can be different. Yet IMHO html field can allow html in twig by default.

**Comment by bobdenotter** [comment]
Sure? I can't reproduce that anymore. 

![schermafbeelding_2019-03-07_om_14_39_50](https://user-images.githubusercontent.com/1833361/53960606-f941b700-40e6-11e9-83de-53625a062621.png)

Agreed on your other points.

**Comment by JarJak** [comment]
So looks like it got fixed  :)

**Comment by xiaohutai** [comment]
I agree with the option to opt-in by setting a `allow_markup`/`allow_html` to `true` (default: `false`) for `text` and `textarea` fields. This is also good information for stripping tags when using it in the context of `<title>` and `<meta>` tags.

**Comment by JarJak** [comment]
Actually I see in Bolt 3 html from title is stripped, not escaped.

**Comment by bobdenotter** [comment]
This was fixed in #354
